File: js/1167.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

const V = +inputs[0];
const tree = Array(V + 1)
  .fill()
  .map(() => []);
const isVisited = Array(V + 1).fill(false);

for (let i = 1; i <= V; i++) {
  const [a, ...rest] = inputs[i].split(' ').map(Number);

  for (let j = 0; j < rest.length - 1; j += 2) {
    const b = rest[j];
    const dist = rest[j + 1];
    tree[a].push([dist, b]);
  }
}

let maxDist = 0;
let endPoint = 0;

isVisited[1] = true;
dfs(1, 0);
initIsVisited();
isVisited[endPoint] = true;
dfs(endPoint, 0);

console.log(maxDist);

function dfs(num, dist) {
  if (dist > maxDist) {
    maxDist = dist;
    maxDistNode = num;
  }
  maxDist = Math.max(maxDist, dist);
  for (let i = 0; i < tree[num].length; i++) {
    const nextNode = tree[num][i][1];
    const nextDist = dist + tree[num][i][0];
    if (!isVisited[nextNode]) {
      isVisited[nextNode] = true;
      dfs(nextNode, nextDist);
    }
  }
}

function initIsVisited() {
  for (let i = 1; i <= V; i++) {
    isVisited[i] = false;
  }
}

File: c++/1167.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<pair<int, int>> adjList[100001];
bool isVisited[100001];
int V;
int maxDist = 0;
int endPoint = 0;

void initIsVisited() {
  for (int i = 1; i <= V; i++) {
    isVisited[i] = false;
  }
}

void dfs(int node, int dist) {
  if (dist > maxDist) {
    maxDist = dist;
    endPoint = node;
  }

  for (int i = 0; i < adjList[node].size(); i++) {
    int nextNode = adjList[node][i].second;
    int nextDist = adjList[node][i].first;

    if (!isVisited[nextNode]) {
      isVisited[nextNode] = true;
      dfs(nextNode, dist + nextDist);
    }
  }
}

int main() {
  freopen("example.txt", "r", stdin);
  cin >> V;

  for (int i = 0; i < V; i++) {
    int a;
    cin >> a;
    while (true) {
      int b, dist;
      cin >> b;
      if (b == -1) break;
      cin >> dist;
      adjList[a].push_back({dist, b});
    }
  }

  isVisited[1] = true;
  dfs(1, 0);

  initIsVisited();
  isVisited[endPoint] = true;
  dfs(endPoint, 0);

  cout << maxDist;
  return 0;
}